src/parser/smt2/smt2.cpp

```diff
@@ -1566,5 +1566,355 @@ InputLanguage Smt2::getLanguage() const
   return em->getOptions().getInputLanguage();
 }
 
+void Smt2::applyTypeAscription(ParseOp& p, Type type)
+{
+  // (as const (Array T1 T2))
+  if (p.d_kind == kind::STORE_ALL)
+  {
+    if (!type.isArray())
+    {
+      std::stringstream ss;
+      ss << "expected array constant term, but cast is not of array type"
+         << std::endl
+         << "cast type: " << type;
+      parseError(ss.str());
+    }
+    p.d_type = type;
+    return;
+  }
+  if (p.d_expr.isNull())
+  {
+    Trace("parser-overloading")
+        << "Getting variable expression with name " << p.d_name << " and type "
+        << type << std::endl;
+    // get the variable expression for the type
+    if (isDeclared(p.d_name, SYM_VARIABLE))
+    {
+      p.d_expr = getExpressionForNameAndType(p.d_name, type);
+    }
+    if (p.d_expr.isNull())
+    {
+      std::stringstream ss;
+      ss << "Could not resolve expression with name " << p.d_name
+         << " and type " << type << std::endl;
+      parseError(ss.str());
+    }
+  }
+  ExprManager* em = getExprManager();
```

Comment:
Just as a note: In smt2.cpp `getExprManager()` is called quite often. It would probably make sense to store a `d_exprmgr` member in `Smt2`.

Reply:
I tried doing this but its a bit tricky, since the expression manager may not be ready or may be changed from the time `Smt2` is created and when we parse.  More generally caching this pointer makes things a bit rigid since it makes the assumption that a single expression manager is used throughout the course of the parser object.

Reply:
Why would we have different expression managers while parsing?

Reply:
Say if we had an architecture where we used a single parser object to parse multiple files in multiple threads.  I'm not 100% sure of the intended architecture, but caching the expression manager in this class would introduce an assumption that we weren't planning on doing that kind of thing.

Reply:
I'm not even sure if the parser object itself can handle multi-threading. I would not want to go there :-) The whole thread-safety/singleton business needs to be revisited anyways.

Reply:
I tend to agree that we probably don't have to worry about multi-threading in that sense. We may want to support parallel parsing but in that case, we'd have likely one parser instance per thread because that seems like the most straightforward architecture. However, in the current version `getExprManager()` is only called three times, which I find reasonable.